File: LeetCode/Python/704.binary-search.py
# @before-stub-for-debug-begin
from python3problem704 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=704 lang=python3
#
# [704] Binary Search
#
# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1 
        mid = 0

        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                right = mid - 1 
            else:
                left = mid + 1


        return -1
# @lc code=end

File: LeetCode/CPP/704.binary-search.cpp
#include <vector>
/*
 * @lc app=leetcode id=704 lang=cpp
 *
 * [704] Binary Search
 */
using namespace std;
// @lc code=start
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int l = 0, r = nums.size();
        int m;
        while (l < r)
        {
            m = l + (r - l) / 2;
            if (nums[m] == target)
            {
                return m;
            }
            else if (nums[m] > target)
            {
                r = m;
            }
            else
            {
                l = m + 1;
            }
        }
        return -1;
    }
};
// @lc code=end